### PR TITLE
Jaccard similarity for paired data

### PR DESCRIPTION
**Pull request by natsheh**
Adding Jaccard similarity as an additional transformer for paired data

### CODE REVIEW

File: beard/similarity/pairs.py [hunk 1]
```diff
@@ -10,13 +10,16 @@
...
+.. codeauthor:: Hussein AL-NATSHEH <h.natsheh@ciapple.com>
 
 """
+from __future__ import division
 import numpy as np
 import scipy.sparse as sp
 
 from sklearn.base import BaseEstimator
 from sklearn.base import TransformerMixin
+from sklearn.metrics import jaccard_similarity_score
```
Comment (glouppe) [inline beard/similarity/pairs.py]: You can remove this, as it is not necessary

File: beard/similarity/tests/test_pairs.py [hunk 2]
```diff
@@ -87,3 +89,22 @@ def test_absolute_difference():
 
     Xt = AbsoluteDifference().fit_transform(sp.csr_matrix(X))
     assert_array_almost_equal(Xt, [[0, 0], [1, 1], [0, 0], [1, 1]])
+
+
+def test_JaccardSimilarity():
+    X = np.array([[0, 1, 0, 1, 0, 1, 0, 1],
+                  [0, 0, 1, 1, 0, 1, 0, 1],
+                  [0, 1, 0, 1, 0, 0, 1, 0],
+                  [1, 0, 1, 1, 1, 1, 0, 1]])
```
Comment (glouppe) [inline beard/similarity/tests/test_pairs.py]: Tests dont pass if you add a row `[0, 0, 0, 0, 0, 0, 0, 0]` and the corresponding `[0.]` output.

File: beard/similarity/pairs.py [hunk 3]
```diff
@@ -273,3 +276,74 @@ def transform(self, X):
...
+            return np.array(Xt).reshape(-1, 1)
+        else:
+            if len(X[X.nonzero()]) == 0.:
+                return np.zeros(n_samples).reshape(-1, 1)
+            numerator = (X1 * X2).sum(axis=1)
+            X_sum = X1 + X2
+            X_sum[X_sum.nonzero()] = 1
+            denominator = X_sum.sum(axis=1)
+            Xt = numerator / denominator
+            return np.array(Xt).reshape(-1, 1)
```
Comment (MSusik) [inline beard/similarity/pairs.py]: Please compare usage of the whitespaces:

`CosineSimilarity`
``` python
        n_samples, n_features_all = X.shape
        n_features = n_features_all // 2
        sparse = sp.issparse(X)

        if sparse and not sp.isspmatrix_csr(X):
            X = X.tocsr()

        X1 = X[:, :n_features]
        X2 = X[:, n_features:]

        if sparse:
            numerator = np.asarray(X1.multiply(X2).sum(axis=1)).ravel()
            norm1 = np.asarray(X1.multiply(X1).sum(axis=1)).ravel()
            norm2 = np.asarray(X2.multiply(X2).sum(axis=1)).ravel()

        else:
            numerator = (X1 * X2).sum(axis=1)
            norm1 = (X1 * X1).sum(axis=1)
            norm2 = (X2 * X2).sum(axis=1)

        denominator = (norm1 ** 0.5) * (norm2 ** 0.5)

        with np.errstate(divide="ignore"):
            Xt = numerator / denominator
            Xt[denominator == 0.0] = 0.0

        return Xt.reshape((n_samples, 1))
```

`Jaccard similarity`
```python
        n_samples, n_features_all = X.shape
        n_features = n_features_all // 2
        X1 = X[:, :n_features]
        X2 = X[:, n_features:]
        sparse = sp.issparse(X)
        if sparse and not sp.isspmatrix_csr(X):
            X = X.tocsr()
        if sparse:
            if X.data[X.data != 0.].sum() == 0:
                return np.zeros(n_samples).reshape(-1, 1)
            numerator = np.asarray(X1.multiply(X2).sum(axis=1)).ravel()
            X_sum = X1 + X2
            X_sum.data[X_sum.data != 0.] = 1
            M = X_sum.sum(axis=1)
            A = M.getA()
            denominator = A.reshape(-1,)
            Xt = numerator / denominator
            return np.array(Xt).reshape(-1, 1)
        else:
            if len(X[X.nonzero()]) == 0.:
                return np.zeros(n_samples).reshape(-1, 1)
            numerator = (X1 * X2).sum(axis=1)
            X_sum = X1 + X2
            X_sum[X_sum.nonzero()] = 1
            denominator = X_sum.sum(axis=1)
            Xt = numerator / denominator
            return np.array(Xt).reshape(-1, 1)

```

IMO the first sample of code is much easier to read.

File: beard/similarity/pairs.py [hunk 4]
```diff
@@ -273,3 +276,86 @@ def transform(self, X):
...
+    ratio between the size of their intersection and the size of their
+    union.
+    """
+
+    def fit(self, X, y=None):
+        """(Do nothing).
+
+        Parameters
+        ----------
+        :param X: array-like, shape (n_samples, n_features) Input data.
```
Comment (glouppe) [inline beard/similarity/pairs.py]: `Input data` should be on a new line

File: beard/similarity/pairs.py [hunk 5]
```diff
@@ -273,3 +276,86 @@ def transform(self, X):
...
+        X2 = X[:, n_features:]
+
+        sparse = sp.issparse(X)
+
+        if sparse and not sp.isspmatrix_csr(X):
+            X = X.tocsr()
+
+        if sparse:
+
+            if X.data[X.data != 0.].sum() == 0:
```
Comment (glouppe) [inline beard/similarity/pairs.py]: Why not simply `X.data.sum() == 0`?

File: beard/similarity/pairs.py [hunk 6]
```diff
@@ -273,3 +276,86 @@ def transform(self, X):
...
+                return np.zeros(n_samples).reshape(-1, 1)
+
+            numerator = np.asarray(X1.multiply(X2).sum(axis=1)).ravel()
```
Comment (glouppe) [inline beard/similarity/pairs.py]: This only work for binary values. I think you should binarize `X` before compute the similarity.

File: beard/similarity/pairs.py [hunk 7]
```diff
@@ -273,3 +276,86 @@ def transform(self, X):
...
+
+        sparse = sp.issparse(X)
+
+        if sparse and not sp.isspmatrix_csr(X):
+            X = X.tocsr()
+
+        if sparse:
+
+            if X.data[X.data != 0.].sum() == 0:
+                return np.zeros(n_samples).reshape(-1, 1)
```
Comment (glouppe) [inline beard/similarity/pairs.py]: You can directly return `np.zeros((n_samples, 1))`

File: beard/similarity/pairs.py [hunk 8]
```diff
@@ -273,3 +276,86 @@ def transform(self, X):
...
+            X_sum = X1 + X2
+            X_sum.data[X_sum.data != 0.] = 1
+            M = X_sum.sum(axis=1)
+            A = M.getA()
+            denominator = A.reshape(-1,)
+
+        else:
+
+            if len(X[X.nonzero()]) == 0.:
+                return np.zeros(n_samples).reshape(-1, 1)
```
Comment (glouppe) [inline beard/similarity/pairs.py]: You can directly return `np.zeros((n_samples, 1))`

File: beard/similarity/pairs.py [hunk 9]
```diff
@@ -273,3 +276,86 @@ def transform(self, X):
...
+
+            numerator = (X1 * X2).sum(axis=1)
+
+            X_sum = X1 + X2
+            X_sum[X_sum.nonzero()] = 1
+            denominator = X_sum.sum(axis=1)
+
+        Xt = numerator / denominator
```
Comment (glouppe) [inline beard/similarity/pairs.py]: Use a with `np.errstate(divide="ignore")` statement around this

File: beard/similarity/tests/test_pairs.py [hunk 10]
```diff
@@ -87,3 +89,29 @@ def test_absolute_difference():
...
+                  [1, 0, 1, 1, 1, 1, 0, 1],
+                  [0, 1, 0, 1, 0, 1, 0, 1]])
+
+    Xt = JaccardSimilarity().fit_transform(X)
+    assert_array_almost_equal(Xt, [[0.], [0.33333333], [0.], [0.5], [1.]])
+
+    Xt = JaccardSimilarity().fit_transform(sp.csr_matrix(X))
+    assert_array_almost_equal(Xt, [[0.], [0.33333333], [0.], [0.5], [1.]])
+
+    X = np.array([[0, 0, 0, 0, 0, 0, 0, 0],
```
Comment (glouppe) [inline beard/similarity/tests/test_pairs.py]: These tests are redundant with the ones above

File: beard/similarity/tests/test_pairs.py [hunk 11]
```diff
@@ -87,3 +89,29 @@ def test_absolute_difference():
 
     Xt = AbsoluteDifference().fit_transform(sp.csr_matrix(X))
     assert_array_almost_equal(Xt, [[0, 0], [1, 1], [0, 0], [1, 1]])
+
+
+def test_JaccardSimilarity():
+    """Test for JaccardSimilarity."""
```
Comment (glouppe) [inline beard/similarity/tests/test_pairs.py]: Can you add tests when `X` is not binary?

**Comment by glouppe** [comment]
Once my comments are addressed, +1 for merge. Please squash all your commits once it is done.

**Pull request closed without merging**